src/phoenix/core/traces.py

```diff
@@ -1,108 +1,288 @@
+            return list(
+                map(
+                    self._spans.__getitem__,
+                    self._traces[trace_id],
+                )
+            )
+
+    def get_spans(
+        self,
+        start_time: Optional[datetime] = None,
+        stop_time: Optional[datetime] = None,
+        root_spans_only: Optional[bool] = False,
+    ) -> List[ReadableSpan]:
+        if not self._spans:
+            return []
+        start_time = start_time or cast(
+            datetime,
+            self._min_start_time,
+        )
+        stop_time = stop_time or (
+            cast(
+                datetime,
+                self._max_start_time,
+            )
+            + timedelta(minutes=1)
+        )
+        sorted_span_ids = (
+            self._start_time_sorted_root_span_ids
+            if root_spans_only
+            else self._start_time_sorted_span_ids
+        )
+        with self._lock:
+            return list(
+                map(
+                    self._spans.__getitem__,
+                    sorted_span_ids.irange_key(
+                        start_time.timestamp(),
+                        stop_time.timestamp(),
+                        inclusive=(True, False),
+                        reverse=True,  # most recent spans first
+                    ),
+                )
+            )
+
+    def latency_rank_percent(
+        self,
+        latency_ms: float,
+    ) -> float:
+        root_span_ids = self._latency_sorted_root_span_ids
+        with self._lock:
+            return cast(
+                int,
+                root_span_ids.bisect_key_left(
+                    latency_ms,
+                ),
+            ) / len(root_span_ids)
+
+    def get_descendant_span_ids(self, span_id: SpanID) -> List[SpanID]:
+        with self._lock:
+            return list(
+                _get_descendant_span_ids(
+                    span_id,
+                    self._child_span_ids,
+                )
+            )
+
+    @property
+    def span_count(self) -> int:
+        """Total number of spans (excluding orphan spans if any)"""
+        with self._lock:
+            return len(self._spans)
+
+    @property
+    def min_start_time(self) -> Optional[datetime]:
+        with self._lock:
+            return self._min_start_time
+
+    @property
+    def max_start_time(self) -> Optional[datetime]:
+        with self._lock:
+            return self._max_start_time
+
+    def __len__(self) -> int:
+        """Total number of traces"""
+        with self._lock:
+            return len(self._start_time_sorted_root_span_ids)
+
+    def __getitem__(self, key: SpanID) -> Optional[ReadableSpan]:
+        with self._lock:
+            return self._spans.get(key)
+
+    def _consume_spans(self) -> None:
+        while True:
+            if not (span := self._queue.get()):
+                return
+            with self._lock:
+                self._process_span(span)
+
+    def _process_span(self, span: Span) -> None:
```

Comment:
Docstring would help here, or breaking into smaller parts that are easier to understand. There's a lot going on in this one.

Reply:
We can refactor this in the future when we have a better understanding. Overhead should be minimized for the first version of this.